README
NAME
    Mojolicious::Plugin::CGI - Run CGI script from Mojolicious

VERSION
    0.35

DESCRIPTION
    This plugin enable Mojolicious to run Perl CGI scripts. It does so by
    forking a new process with a modified environment and reads the STDOUT
    in a non-blocking manner.

SYNOPSIS
  Standard usage
      use Mojolicious::Lite;
      plugin CGI => [ "/cgi-bin/script" => "/path/to/cgi/script.pl" ];

    Using the code above is enough to run "script.pl" when accessing
    <http://localhost:3000/cgi-bin/script>.

  Complex usage
      plugin CGI => {
        # Specify the script and mount point
        script => "/path/to/cgi/script.pl",
        route  => "/some/route",

        # %ENV variables visible from inside the CGI script
        env => {}, # default is \%ENV

        # Path to where STDERR from cgi script goes
        errlog => "/path/to/file.log",

        # The "before" hook is called before script start
        # It receives a Mojolicious::Controller which can be modified
        before => sub {
          my $c = shift;
          $c->req->url->query->param(a => 123);
        },
      };

    The above contains all the options you can pass on to the plugin.

  Helper
      plugin "CGI";

      # GET /cgi-bin/some-script.cgi/path/info?x=123
      get "/cgi-bin/#script_name/*path_info" => {path_info => ''}, sub {
        my $c    = shift;
        my $name = $c->stash("script_name");
        $c->cgi->run(script => File::Spec->rel2abs("/path/to/cgi/$name"));
      };

    The helper can take most the arguments that "register" takes, with the
    exception of "support_semicolon_in_query_string".

    It is critical that "script_name" and "path_info" is present in stash.
    If the values are extracted directly from the path or set manually does
    not matter.

    Note that the helper is registered in all of the examples.

  Running code refs
      plugin CGI => {
        route => "/some/path",
        run   => sub {
          my $cgi = CGI->new;
          # ...
        }
      };

    Instead of calling a script, you can run a code block when accessing the
    route. This is (pretty much) safe, even if the code block modifies
    global state, since it runs in a separate fork/process.

  Support for semicolon in query string
      plugin CGI => {
        support_semicolon_in_query_string => 1,
        ...
      };

    The code above need to be added before other plugins or handler which
    use "url" in Mojo::Message::Request. It will inject a "before_dispatch"
    hook which saves the original QUERY_STRING, before it is split on "&" in
    Mojo::Parameters.

ATTRIBUTES
  env
    Holds a hash ref containing the environment variables that should be
    used when starting the CGI script. Defaults to %ENV when this module was
    loaded.

    This plugin will create a set of environment variables depenendent on
    the request passed in which is according to the CGI spec. In addition to
    "env", these dynamic variables are set:

      CONTENT_LENGTH, CONTENT_TYPE, HTTPS, PATH, PATH_INFO, QUERY_STRING,
      REMOTE_ADDR, REMOTE_HOST, REMOTE_PORT, REMOTE_USER, REQUEST_METHOD,
      SCRIPT_NAME, SERVER_PORT, SERVER_PROTOCOL.

    Additional static variables:

      GATEWAY_INTERFACE = "CGI/1.1"
      SERVER_ADMIN = $ENV{USER}
      SCRIPT_FILENAME = Script name given as argument to register.
      SERVER_NAME = Sys::Hostname::hostname()
      SERVER_SOFTWARE = "Mojolicious::Plugin::CGI"

    Plus all headers are exposed. Examples:

      .----------------------------------------.
      | Header          | Variable             |
      |-----------------|----------------------|
      | Referer         | HTTP_REFERER         |
      | User-Agent      | HTTP_USER_AGENT      |
      | X-Forwarded-For | HTTP_X_FORWARDED_FOR |
      '----------------------------------------'

  register
      $self->register($app, [ $route => $script ]);
      $self->register($app, %args);
      $self->register($app, \%args);

    "route" and path need to exist as keys in %args unless given as plain
    arguments.

    $route can be either a plain path or a route object.

COPYRIGHT AND LICENSE
    Copyright (C) 2014, Jan Henning Thorsen

    This program is free software, you can redistribute it and/or modify it
    under the terms of the Artistic License version 2.0.

AUTHOR
    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Revision history for perl distribution Mojolicious-Plugin-CGI

0.40 Not Released
 - Compatible with Mojolicious 8.06 #30

0.39 2018-09-02T22:33:56+0200
 - Compatible with Mojolocious 7.90 #29

0.38 2017-06-12T21:05:08+0200
 - Bumped Mojolicious version to 7.28 #27

0.37 2017-05-15T22:01:59+0200
 - Stop using deprecated "spurt" #26

0.36 2016-10-27T11:55:46+0200
 - Fix setting $ENV{PATH_INFO} with UTF-8 characters
   Contributor: kensanata

0.35 2016-10-13T10:04:50+0200
 - Skipping tests on Windows

0.34 2016-06-10T19:41:37-0700
 - zombies.t cannot run without File::Which

0.33 2016-06-07T15:46:09-0500
 - Fix zombies left behind #20 #21 #22

0.32 2016-04-17T14:34:17+0200
 - Fix passing on %ENV to cgi script #19

0.31 2016-04-11T18:17:20+0200
 - Avoid making CPAN dist with files from t/cgi-bin

0.30 2016-04-11T18:06:04+0200
 - Will generate all the cgi-scripts with $^X

0.29 2016-04-09T10:54:27+0200
 - Replace "/usr/bin/perl" with "/usr/bin/env perl" in tests

0.28 2016-04-06T16:22:50+0200
 - Try to fix "Could not run CGI script" test failures

0.27 2016-04-01T14:09:59+0200
 - Refactored the module
 - Add $c->cgi->run(...) helper
 - Fix bug when exec() fail in child
 - Remove "ioloop" attribute
 - Remove "emulate_environment" method

0.26 2016-03-10T22:15:51+0100
 - Expose all headers as environment variables #18

0.25 2015-11-27T11:49:54+0100
 - Made post.t optional

0.24 2015-11-11T22:16:33+0100
 - Add "run" option for running CGI code from a callback instead of script #17
 - Add HTTP_IF_NONE_MATCH to environement #17
 - Add support for reading the "Status" header
 - Improved NPH script support #17

0.23 2015-07-16T07:43:59+0200
 - Try to fix waiting for finished CGI scripts, without checking
   return value from waitpid()

0.22 2015-07-14T17:32:59+0200
 - Fix "Bareword found where operator expected at t/gh-16-loop_reset.t line 35"
 - Try to fix waiting for finished CGI scripts

0.21 2015-07-12T15:26:05+0200
 - Fix "Can't call method "write" on an undefined value" #16
 - Disabled "no leaky leaks" test

0.20 2015-03-13T08:27:48Z
 - t/post.t leaky pipes test works on osx (timgimyee)

0.19 2015-02-08T14:24:43Z
 - Will not test SERVER_ADMIN variable

0.18 2015-01-30T09:57:46Z
 - Fix SCRIPT_NAME environment variable
 - Fix minimum Mojolicious version in cpanfile #13

0.17 2015-01-15T17:19:57Z
 - Add support for QUERY_STRING with ";" instead of "&"

0.16 2015-01-02T13:24:59Z
 - Require Mojolicious 5.08

0.15 2014-12-14T15:24:14Z
 - Fix CPAN testers and HTTP_HOST regexp

0.14 2014-12-12T13:27:17Z
 - Fix multipart test content length value Contributor: leejo

0.13 2014-11-17T15:13:31Z
 - Reset IOLoop in fork to prevent fork from receiving connections

0.12 2014-10-05T18:44:58Z
 - Change default CGI logging prefix to "CGI:$name:$pid"
 - Fix _waitpids() actually checks hanging processes

0.11 2014-10-05T10:58:12Z
 - Default STDERR logging goes to app log
 - Refactored the code using IO::Pipely
 - env() is an attribute
 - Tried to fix REMOTE_HOST from Authorization header, closes #10

0.10 2014-10-02T21:31:18Z
 - Fix Can't use string ("") as a HASH ref while "strict refs" in use at
   Mojolicious::Routes::Pattern line 63.

0.09 2014-06-13T11:56:31Z
 - Add README.pod to MANIFEST.SKIP
   https://github.com/jhthorsen/mojolicious-plugin-cgi/issues/9

0.08 2014-05-14T10:29:24Z
 - Flushing STDERR as well
 - Fix REMOTE_HOST test on windows
   http://www.cpantesters.org/cpan/report/99cac79a-725d-1014-82ca-1dcb5cf4d
   ae8 REMOTE_HOST=599DSOLIMANO01.cowen.corp != localhost

0.07 2013-11-17T14:11:59Z
 - Reads userinfo from request path
 - Add support for errlog
 - Fix PATH_INFO and SCRIPT_NAME

0.0601 2013-10-01T12:36:25Z
 - Add repository to Makefile.PL

0.06 2013-09-24T09:43:37Z
 - Add support multipart requests lee@givengain.ch

0.0501 2013-08-26T18:35:24Z
 - Fix waitpid before kill, because of defunc children

0.05 2013-08-23T16:31:13Z
 - Fix pipe leak
 - Support 302 redirects
 - Add better error handling
 - Will check if child is running

0.0401 2013-08-16T23:24:58Z
 - Fix NAME in Makefile.PL

0.04 2013-08-16T09:15:16Z
 - Add before hook which can modify query string

0.0301 2013-08-08T12:31:28Z
 - Need to waitpid to avoid defunct processes

0.03 2013-08-08T12:23:36Z
 - CGI script is now non-blocking

0.02 2013-08-05T12:45:43Z
 - Fix compat with older perls.

0.01 2013-07-11T11:40:26Z
 - Add basic support for calling CGI scripts



t/Helper.pm
package t::Helper;
use Mojo::Base -strict;

use File::Basename;
use File::Spec::Functions qw(catdir catfile);
use Mojolicious;
use Mojo::Loader;
use Test::Mojo;
use Test::More;

sub cgi_script {
  my $template = shift;
  my $script = catfile 't', 'cgi-bin', $template;
  mkdir catdir qw(t cgi-bin);
  open my $CGI_BIN, '>', $script or Test::More::plan(skip_all => "write $script: $!");
  print $CGI_BIN "#!$^X\n";
  print $CGI_BIN "use strict;\nuse warnings;\n";
  print $CGI_BIN Mojo::Loader::data_section(__PACKAGE__, $template);
  close $CGI_BIN;
  eval { chmod 0755, $script };
  return $script;
}

sub import {
  my $class  = shift;
  my $caller = caller;

  eval <<"HERE";
package $caller;
use Mojo::Base -strict;
use Test::Mojo;
use Test::More;
HERE

  Mojo::Util::monkey_patch($caller => cgi_script => \&cgi_script);
}

1;

__DATA__
@@ basic.pl
print "Content-Type: text/custom\n\r\n\rbasic stuff\n";
@@ env
print "Content-Type: text/plain\n\r";
print "\n\rENVIRON";
print "MENT\n";
print "$_=$ENV{$_}\n" for sort keys %ENV;
@@ env.cgi
print "Content-Type: text/plain\n\r";
print "\n\rENVIRON";
print "MENT\n";
print "$_=$ENV{$_}\n" for sort keys %ENV;
@@ errlog
warn "yikes!";
print "Content-Type: text/plain\n\r\n\r";
print "yayayyaya\n";
@@ file_upload
print "Content-Type: text/custom\n\r\n\r";
print "$$\n";
print "=== $ENV{$_}\n" for qw/CONTENT_TYPE CONTENT_LENGTH/;
print "--- $_" while <STDIN>;
@@ not-found.pl
print "Status: 404 Not Found\r\n";
print "Content-Type: text/html; charset=ISO-8859-1\r\n";
print "\r\n";
print "<body><p>This page is missing\n";
@@ not-modified.pl
print "Status: 304 Not Modified\r\n";
print "X-Test: if-none-match seen: $ENV{HTTP_IF_NONE_MATCH}\r\n";
print "\r\n";
@@ nph-borked.pl
# When SERVER_PROTOCOL is set to "HTTP", the CGI module will just print HTTP and
# no version!
print "HTTP 403 Payment Required\r\n";
print "Content-Type: text/html; charset=ISO-8859-1\r\n";
print "\r\n";
print "<body><p>This is the borked paywall.\n";
@@ nph.pl
print "HTTP/1.1 403 Payment Required\r\n";
print "Content-Type: text/html; charset=ISO-8859-1\r\n";
print "\r\n";
print "<body><p>This is the paywall.\n";
@@ postman
print "Content-Type: text/custom\n\r\n\r";
print "$$\n";
print "--- $_" while <STDIN>;
@@ redirect.pl
print "Location: http://somewhereelse.com\n\r\n\r";
@@ slow.pl
sleep 1;
print "Content-Type: text/custom\n\r\n\rHello Morbo!\n";


t/cgi-bin/nph.pl
#!/usr/bin/env perl
use strict;
use warnings;

print "HTTP/1.1 403 Payment Required\r\n";
print "Content-Type: text/html; charset=ISO-8859-1\r\n";
print "\r\n";
print "<body><p>This is the paywall.\n";


t/cgi-bin/slow.pl
#!/usr/bin/env perl
use strict;
use warnings;
sleep 1;
print "Content-Type: text/custom\n\r\n\rHello Morbo!\n";


Makefile.PL
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
use ExtUtils::MakeMaker;
WriteMakefile(
  NAME          => 'Mojolicious::Plugin::CGI',
  AUTHOR        => 'Jan Henning Thorsen <jhthorsen@cpan.org>',
  LICENSE       => 'artistic_2',
  ABSTRACT_FROM => 'lib/Mojolicious/Plugin/CGI.pm',
  VERSION_FROM  => 'lib/Mojolicious/Plugin/CGI.pm',
  EXE_FILES     => [qw(  )],
  META_MERGE    => {
    resources => {
      bugtracker => 'https://github.com/jhthorsen/mojolicious-plugin-cgi/issues',
      homepage   => 'https://github.com/jhthorsen/mojolicious-plugin-cgi',
      repository => 'https://github.com/jhthorsen/mojolicious-plugin-cgi.git',
    },
  },
  BUILD_REQUIRES => {'Test::More' => '0.88'},
  PREREQ_PM      => {'IO::Pipely' => '0.005', 'Mojolicious' => '6.49'},
  test           => {TESTS        => (-e 'META.yml' ? 't/*.t' : 't/*.t xt/*.t')},
);


lib/Mojolicious/Plugin/CGI.pm
package Mojolicious::Plugin::CGI;
use Mojo::Base 'Mojolicious::Plugin';

use File::Basename;
use File::Spec;
use IO::Pipely 'pipely';
use Mojo::Util qw(b64_decode encode);
use POSIX 'WNOHANG';
use Perl::OSType 'is_os_type';
use Socket qw(AF_INET inet_aton);
use Sys::Hostname;

use constant CHECK_CHILD_INTERVAL => $ENV{CHECK_CHILD_INTERVAL} || 0.01;
use constant DEBUG                => $ENV{MOJO_PLUGIN_CGI_DEBUG};
use constant IS_WINDOWS           => is_os_type('Windows');
use constant READ                 => 0;
use constant WRITE                => 1;

our $VERSION      = '0.40';
our %ORIGINAL_ENV = %ENV;

has env => sub { +{%ORIGINAL_ENV} };

sub register {
  my ($self, $app, $args) = @_;
  my $pids = $app->{'mojolicious_plugin_cgi.pids'} ||= {};

  $args = {route => shift @$args, script => shift @$args} if ref $args eq 'ARRAY';
  $args->{env} ||= $self->env;
  $args->{run} = delete $args->{script} if ref $args->{script} eq 'CODE';
  $args->{pids} = $pids;

  $app->helper('cgi.run' => sub { _run($args, @_) }) unless $app->renderer->helpers->{'cgi.run'};
  $app->{'mojolicious_plugin_cgi.tid'}
    ||= Mojo::IOLoop->recurring(CHECK_CHILD_INTERVAL, sub { local ($?, $!); _waitpids($pids); });

  if ($args->{support_semicolon_in_query_string}
    and !$app->{'mojolicious_plugin_cgi.before_dispatch'}++)
  {
    $app->hook(
      before_dispatch => sub {
        $_[0]->stash('cgi.query_string' => $_[0]->req->url->query->to_string);
      }
    );
  }

  return unless $args->{route};    # just register the helper
  die "Neither 'run', nor 'script' is specified." unless $args->{run} or $args->{script};
  $args->{route} = $app->routes->any("$args->{route}/*path_info", {path_info => ''})
    unless ref $args->{route};
  $args->{script} = File::Spec->rel2abs($args->{script}) || $args->{script} if $args->{script};
  $args->{route}->to(cb => sub { _run($args, @_) });
}

sub _child {
  my ($c, $args, $stdin, $stdout, $stderr) = @_;
  my @STDERR = @$stderr ? ('>&', fileno $stderr->[WRITE]) : ('>>', $args->{errlog});

  Mojo::IOLoop->reset;
  warn "[CGI:$args->{name}:$$] <<< (@{[$stdin->slurp]})\n" if DEBUG;
  open STDIN, '<', $stdin->path or die "STDIN @{[$stdin->path]}: $!" if -s $stdin->path;
  open STDERR, $STDERR[0], $STDERR[1] or die "STDERR: @$stderr: $!";
  open STDOUT, '>&', fileno $stdout->[WRITE] or die "STDOUT: $!";
  select STDERR;
  $| = 1;
  select STDOUT;
  $| = 1;

  %ENV = _emulate_environment($c, $args);
  $args->{run} ? $args->{run}->($c) : exec $args->{script}
    || die "Could not execute $args->{script}: $!";

  eval { POSIX::_exit($!) } unless IS_WINDOWS;
  eval { CORE::kill KILL => $$ };
  exit $!;
}

sub _emulate_environment {
  my ($c, $args) = @_;
  my $tx             = $c->tx;
  my $req            = $tx->req;
  my $headers        = $req->headers;
  my $content_length = $req->content->is_multipart ? $req->body_size : $headers->content_length;
  my %env_headers    = (HTTP_COOKIE => '', HTTP_REFERER => '');
  my ($remote_user, $script_name);

  for my $name (@{$headers->names}) {
    my $key = uc "http_$name";
    $key =~ s!\W!_!g;
    $env_headers{$key} = $headers->header($name);
  }

  if (my $userinfo = $c->req->url->to_abs->userinfo) {
    $remote_user = $userinfo =~ /([^:]+)/ ? $1 : '';
  }
  elsif (my $authenticate = $headers->authorization) {
    $remote_user = $authenticate =~ /Basic\s+(.*)/ ? b64_decode $1 : '';
    $remote_user = $remote_user =~ /([^:]+)/       ? $1            : '';
  }

  if ($args->{route}) {
    $script_name = $c->url_for($args->{route}->name, {path_info => ''})->path->to_string;
  }
  elsif (my $name = $c->stash('script_name')) {
    my $name = quotemeta $name;
    $script_name = $c->req->url->path =~ m!^(.*?/$name)! ? $1 : $c->stash('script_name');
  }

  return (
    %{$args->{env}},
    CONTENT_LENGTH => $content_length        || 0,
    CONTENT_TYPE   => $headers->content_type || '',
    GATEWAY_INTERFACE => 'CGI/1.1',
    HTTPS             => $req->is_secure ? 'YES' : 'NO',
    %env_headers,
    PATH_INFO       => '/' . encode('UTF-8', $c->stash('path_info') // ''),
    QUERY_STRING    => $c->stash('cgi.query_string') || $req->url->query->to_string,
    REMOTE_ADDR     => $tx->remote_address,
    REMOTE_HOST     => gethostbyaddr(inet_aton($tx->remote_address || '127.0.0.1'), AF_INET) || '',
    REMOTE_PORT     => $tx->remote_port,
    REMOTE_USER     => $remote_user || '',
    REQUEST_METHOD  => $req->method,
    SCRIPT_FILENAME => $args->{script} || '',
    SCRIPT_NAME     => $script_name || $args->{name},
    SERVER_ADMIN    => $ENV{USER} || '',
    SERVER_NAME     => hostname,
    SERVER_PORT     => $tx->local_port,
    SERVER_PROTOCOL => $req->is_secure ? 'HTTPS' : 'HTTP',    # TODO: Version is missing
    SERVER_SOFTWARE => __PACKAGE__,
  );
}

sub _run {
  my ($defaults, $c) = (shift, shift);
  my $args = @_ ? @_ > 1 ? {@_} : {%{$_[0]}} : {};
  my $before = $args->{before} || $defaults->{before};
  my $stdin  = _stdin($c);
  my @stdout = pipely;
  my ($pid, $log_key, @stderr);

  $args->{$_} ||= $defaults->{$_} for qw(env errlog route run script);
  $args->{name} = $args->{run} ? "$args->{run}" : basename $args->{script};
  $c->$before($args) if $before;
  @stderr = (pipely) unless $args->{errlog};
  defined($pid = fork) or die "[CGI:$args->{name}] fork failed: $!";
  _child($c, $args, $stdin, \@stdout, \@stderr) unless $pid;
  $args->{pids}{$pid} = $args->{name};
  $log_key = "CGI:$args->{name}:$pid";
  $c->app->log->debug("[$log_key] START @{[$args->{script} || $args->{run}]}");

  for my $p (\@stdout, \@stderr) {
    next unless $p->[READ];
    close $p->[WRITE];
    $p->[READ] = Mojo::IOLoop::Stream->new($p->[READ])->timeout(0);
    Mojo::IOLoop->stream($p->[READ]);
  }

  $c->stash('cgi.pid' => $pid, 'cgi.stdin' => $stdin);
  $c->render_later;

  $stderr[READ]->on(read => _stderr_cb($c, $log_key)) if $stderr[READ];
  $stdout[READ]->on(read => _stdout_cb($c, $log_key));
  $stdout[READ]->on(
    close => sub {
      my $GUARD = 50;
      warn "[CGI:$args->{name}:$pid] Child closed STDOUT\n" if DEBUG;
      unlink $stdin->path or die "Could not remove STDIN @{[$stdin->path]}" if -e $stdin->path;
      local ($?, $!);
      _waitpids({$pid => $args->{pids}{$pid}})
        while $args->{pids}{$pid}
        and kill 0, $pid
        and $GUARD--;
      $defaults->{pids}{$pid} = $args->{pids}{$pid} if kill 0, $pid;
      return $c->finish if $c->res->code;
      return $c->render(text => "Could not run CGI script ($?, $!).\n", status => 500);
    }
  );
}

sub _stderr_cb {
  my ($c, $log_key) = @_;
  my $log = $c->app->log;
  my $buf = '';

  return sub {
    my ($stream, $chunk) = @_;
    warn "[$log_key] !!! ($chunk)\n" if DEBUG;
    $buf .= $chunk;
    $log->warn("[$log_key] $1") while $buf =~ s!^(.+)[\r\n]+$!!m;
  };
}

sub _stdout_cb {
  my ($c, $log_key) = @_;
  my $buf = '';
  my $headers;

  return sub {
    my ($stream, $chunk) = @_;
    warn "[$log_key] >>> ($chunk)\n" if DEBUG;

    # true if HTTP header has been written to client
    return $c->write($chunk) if $headers;

    $buf .= $chunk;

    # false until all headers has been read from the CGI script
    $buf =~ s/^(.*?\x0a\x0d?\x0a\x0d?)//s or return;
    $headers = $1;

    if ($headers =~ /^HTTP/) {
      $c->res->code($headers =~ m!^HTTP (\d\d\d)! ? $1 : 200);
      $c->res->parse($headers);
    }
    else {
      $c->res->code($1) if $headers =~ /^Status: (\d\d\d)/m;
      $c->res->code($headers =~ /Location:/ ? 302 : 200) unless $c->res->code;
      $c->res->parse($c->res->get_start_line_chunk(0) . $headers);
    }
    $c->write($buf) if length $buf;
  };
}

sub _stdin {
  my $c = shift;
  my $stdin;

  if ($c->req->content->is_multipart) {
    $stdin = Mojo::Asset::File->new;
    $stdin->add_chunk($c->req->build_body);
  }
  else {
    $stdin = $c->req->content->asset;
  }

  return $stdin if $stdin->isa('Mojo::Asset::File');
  return Mojo::Asset::File->new->add_chunk($stdin->slurp);
}

sub _waitpids {
  my $pids = shift;

  for my $pid (keys %$pids) {

    # no idea why i need to do this, but it seems like waitpid() below return -1 if not
    local $SIG{CHLD} = 'DEFAULT';
    next unless waitpid $pid, WNOHANG;
    my $name = delete $pids->{$pid} || 'unknown';
    my ($exit_value, $signal) = ($? >> 8, $? & 127);
    warn "[CGI:$name:$pid] Child exit_value=$exit_value ($signal)\n" if DEBUG;
  }
}

1;

=encoding utf8

=head1 NAME

Mojolicious::Plugin::CGI - Run CGI script from Mojolicious

=head1 VERSION

0.40

=head1 DESCRIPTION

This plugin enables L<Mojolicious> to run Perl CGI scripts. It does so by forking
a new process with a modified environment and reads the STDOUT in a non-blocking
manner.

=head1 SYNOPSIS

=head2 Standard usage

  use Mojolicious::Lite;
  plugin CGI => [ "/cgi-bin/script" => "/path/to/cgi/script.pl" ];

Using the code above is enough to run C<script.pl> when accessing
L<http://localhost:3000/cgi-bin/script>.

=head2 Complex usage

  plugin CGI => {
    # Specify the script and mount point
    script => "/path/to/cgi/script.pl",
    route  => "/some/route",

    # %ENV variables visible from inside the CGI script
    env => {}, # default is \%ENV

    # Path to where STDERR from cgi script goes
    errlog => "/path/to/file.log",

    # The "before" hook is called before script start
    # It receives a Mojolicious::Controller which can be modified
    before => sub {
      my $c = shift;
      $c->req->url->query->param(a => 123);
    },
  };

The above contains all the options you can pass on to the plugin.

=head2 Helper

  plugin "CGI";

  # GET /cgi-bin/some-script.cgi/path/info?x=123
  get "/cgi-bin/#script_name/*path_info" => {path_info => ''}, sub {
    my $c    = shift;
    my $name = $c->stash("script_name");
    $c->cgi->run(script => File::Spec->rel2abs("/path/to/cgi/$name"));
  };

The helper can take most of the arguments that L</register> takes, with the
exception of C<support_semicolon_in_query_string>.

It is critical that "script_name" and "path_info" is present in
L<stash|Mojolicious::Controller/stash>. Whether the values are extracted directly
from the path or set manually does not matter.

Note that the helper is registered in all of the examples.

=head2 Running code refs

  plugin CGI => {
    route => "/some/path",
    run   => sub {
      my $cgi = CGI->new;
      # ...
    }
  };

Instead of calling a script, you can run a code block when accessing the route.
This is (pretty much) safe, even if the code block modifies global state,
since it runs in a separate fork/process.

=head2 Support for semicolon in query string

  plugin CGI => {
    support_semicolon_in_query_string => 1,
    ...
  };

The code above needs to be added before other plugins or handlers which use
L<Mojo::Message::Request/url>. It will inject a C<before_dispatch>
hook which saves the original QUERY_STRING, before it is split on
"&" in L<Mojo::Parameters>.

=head1 ATTRIBUTES

=head2 env

Holds a hash ref containing the environment variables that should be
used when starting the CGI script. Defaults to C<%ENV> when this module
was loaded.

This plugin will create a set of environment variables depenendent on the
request passed in which is according to the CGI spec. In addition to L</env>,
these dynamic variables are set:

  CONTENT_LENGTH, CONTENT_TYPE, HTTPS, PATH, PATH_INFO, QUERY_STRING,
  REMOTE_ADDR, REMOTE_HOST, REMOTE_PORT, REMOTE_USER, REQUEST_METHOD,
  SCRIPT_NAME, SERVER_PORT, SERVER_PROTOCOL.

Additional static variables:

  GATEWAY_INTERFACE = "CGI/1.1"
  SERVER_ADMIN = $ENV{USER}
  SCRIPT_FILENAME = Script name given as argument to register.
  SERVER_NAME = Sys::Hostname::hostname()
  SERVER_SOFTWARE = "Mojolicious::Plugin::CGI"

Plus all headers are exposed. Examples:

  .----------------------------------------.
  | Header          | Variable             |
  |-----------------|----------------------|
  | Referer         | HTTP_REFERER         |
  | User-Agent      | HTTP_USER_AGENT      |
  | X-Forwarded-For | HTTP_X_FORWARDED_FOR |
  '----------------------------------------'

=head2 register

  $self->register($app, [ $route => $script ]);
  $self->register($app, %args);
  $self->register($app, \%args);

C<route> and L<path> need to exist as keys in C<%args> unless given as plain
arguments.

C<$route> can be either a plain path or a route object.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014, Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut
